AK+LibCore: Fix compilation errors on Windows

stasoid opened the pull request:
Commits authored by cammo1123 are cherry-picked from [WindowsJS2.0](https://github.com/cammo1123/serenity/tree/WindowsJS2.0) with
`git --git-dir=..\serenity-cammo1123\.git format-patch -k -1 --stdout <commit hash> | git am -3 -k` as described [here](https://stackoverflow.com/a/9507417).

tcl3 commented:
When trying to build this on Linux I get the following error:
```
> ./Meta/ladybird.sh run                                                                                                                                                                                                                   
Checking for CMake version at least 3.25... ok, found 3.28.3
Preset CMake variables:

  CMAKE_BUILD_TYPE="RelWithDebInfo"
  CMAKE_TOOLCHAIN_FILE="/home/tim/repos/ladybird/Toolchain/Tarballs/vcpkg/scripts/buildsystems/vcpkg.cmake"
  SERENITY_CACHE_DIR="/home/tim/repos/ladybird/Build/caches"
  VCPKG_INSTALL_OPTIONS="--no-print-usage"
  VCPKG_OVERLAY_TRIPLETS="/home/tim/repos/ladybird/Meta/CMake/vcpkg/release-triplets"

Preset environment variables:

  LADYBIRD_SOURCE_DIR="/home/tim/repos/ladybird"
  VCPKG_BINARY_SOURCES="clear;files,/home/tim/repos/ladybird/Toolchain/Build/vcpkg-binary-cache,readwrite"
  VCPKG_ROOT="/home/tim/repos/ladybird/Toolchain/Tarballs/vcpkg"

-- Running vcpkg install
Detecting compiler hash for triplet x64-linux...
Compiler found: /usr/bin/c++
All requested packages are currently installed.
Total install time: 210 ns
-- Running vcpkg install - done
-- Found the following ICU libraries:
--   data (required): /home/tim/repos/ladybird/Build/ladybird/vcpkg_installed/x64-linux/lib/libicudata.a
--   i18n (required): /home/tim/repos/ladybird/Build/ladybird/vcpkg_installed/x64-linux/lib/libicui18n.a
--   uc (required): /home/tim/repos/ladybird/Build/ladybird/vcpkg_installed/x64-linux/lib/libicuuc.a
-- Found ZLIB: /home/tim/repos/ladybird/Build/ladybird/vcpkg_installed/x64-linux/lib/libz.a (found suitable version "1.3.1", minimum required is "1")
CMake Warning at Build/ladybird/vcpkg_installed/x64-linux/share/curl/vcpkg-cmake-wrapper.cmake:49 (message):
  CURL_LIBRARIES list at least one target.  This will not work for use cases
  where targets are not resolved.
Call Stack (most recent call first):
  Toolchain/Tarballs/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  Ladybird/RequestServer/CMakeLists.txt:23 (find_package)


-- Found ZLIB: /home/tim/repos/ladybird/Build/ladybird/vcpkg_installed/x64-linux/lib/libz.a (found version "1.3.1")
-- Configuring done (1.8s)
CMake Error at Toolchain/Tarballs/vcpkg/scripts/buildsystems/vcpkg.cmake:639 (_add_library):
  Cannot find source file:

    ProcessStatisticsReader.cpp
Call Stack (most recent call first):
  Meta/Lagom/CMakeLists.txt:217 (add_library)
  Meta/Lagom/CMakeLists.txt:302 (lagom_lib)
  Userland/Libraries/LibCore/CMakeLists.txt:112 (serenity_lib)


CMake Error at Toolchain/Tarballs/vcpkg/scripts/buildsystems/vcpkg.cmake:639 (_add_library):
  No SOURCES given to target: LibCore
Call Stack (most recent call first):
  Meta/Lagom/CMakeLists.txt:217 (add_library)
  Meta/Lagom/CMakeLists.txt:302 (lagom_lib)
  Userland/Libraries/LibCore/CMakeLists.txt:112 (serenity_lib)
```
There also seems to be a linking issue with [the PR linter](https://github.com/LadybirdBrowser/ladybird/actions/runs/11329272250/job/31504427361?pr=1794).

stasoid commented:
My builds on WSL2  are very slow, that's why I wanted to use CI to see if it compiles on Linux.

stasoid commented:
Oh, I didn't realize it will ask for approval on every update. I guess I'll have to build locally on Linux anyway...

AtkinsSJ commented:
> Oh, I didn't realize it will ask for approval on every update. I guess I'll have to build locally on Linux anyway...

It will for your first PR, yeah. :sweat_smile:

stasoid commented:
I think I fixed formatting issues; also squashed some related commits. CONTRIBUTING.md says "Make sure your commits are rebased on the master branch" - I don't know how to do this. Is this needed if there are no conflicts?

AtkinsSJ commented:
> CONTRIBUTING.md says "Make sure your commits are rebased on the master branch" - I don't know how to do this. Is this needed if there are no conflicts?

It's only necessary if you have conflicts, or your branch is really old, which might cause issues even if there isn't an obvious conflict.

The way to do it is to bring your master branch up to date, check out your branch, and then `git rebase master`.

stasoid commented:
Thank you. I didn't quite understand what you said, but luckily that's a problem for another time.

ADKaster reviewed (commented):
Thanks for reviving these old commits. However, there's a lot that needs fixed up before these are mergeable. I realize that you are not the initial author for most of these commits, but as you opened the PR, you've volunteered to be responsible for them. If this feedback is too much to address, that's fine, creating such a large, low-level PR for your first PR is not something I would recommend to most people.

My recommendation would be to split this PR into several byte-sized chunks that can be reviewed independently.

The big themes are that we should be following this guidance: https://github.com/LadybirdBrowser/ladybird/blob/c4f7361a5b7b1bbcbb1aa5a2bd7a31f6d8e5f575/Documentation/Porting.md#windows

We should be avoiding platform ifdefs in headers at great cost

We should be using platform-specific files instead of putting an ifdef in every function

We should not try to keep older commits "pristine". If they need fixups, fix them up.

We should avoid including windows.h from header files at any cost. it is simply too large to include in every translation unit. Only .cpp files should be allowed to include it, and only if they really need it.

nico commented on AK/Platform.h:
```diff
@@ -141,6 +141,9 @@
 #if defined(_WIN32) || defined(_WIN64)
+#    if defined(_MSC_VER)
+#        define AK_COMPILER_MSVC
+#    endif
```
clang-cl sets it too. But yes, it's reduced with AK_OS_WIN

ADKaster commented:
> When I split the PR, what happens to review comments on files that no longer exist in the branch? Do they get lost?

 You should create a different git branch on your fork for each pull request, and leave your master branch alone, as it's associated with this pull request id. The comments aren't going anywhere, they're on GitHub.

Platform ifdefs to decide which headers to include are fine I'm headers. But platform ifdefs in the body of classes is ugly and shows that whatever abstraction you were trying to create is leaky. LibCore for example should hide platform specifics from the caller. If higher level libraries like LibWeb or LibJS need to add ifdefs to call the right method on your class, then the LibCore class or AK class didn't do a good job of being generic.

stasoid commented on Meta/CMake/lagom_compile_options.cmake:
```diff
@@ -46,10 +46,12 @@ function(add_cxx_linker_flag_if_supported flag)
 endfunction()
 
-add_cxx_linker_flag_if_supported(LINKER:--gdb-index)
+if (NOT WIN32)
```
This change fixes these warnings:
```
lld-link : warning : ignoring unknown argument '--gdb-index' 
lld-link : warning : ignoring unknown argument '-Bsymbolic-non-weak-functions'
```

ADKaster replied:
hm that's suprising. makes me think our "add if supported" check is a bit broken. but fair enough.

stasoid commented:
I am trying to understand what to do with DirectoryEntry.h on Windows.

* `inode_number` and `posix_name_from_directory_entry_type` do not seem to be used anywhere, can I remove them?
* to avoid #ifdefs in `DirectoryEntry`, I want to move `from_dirent` and `from_stat` to DirIterator.cpp and `from_find_data` to DirIteratorWindows.cpp. And name them `directory_entry_from_dirent`, etc.  Is it an appropriate solution?

stasoid commented on Userland/Libraries/LibCore/Environment.cpp:
```diff
@@ -161,6 +193,16 @@ ErrorOr<void> clear()
+        auto eq = strchr(str, '=');
+        VERIFY(eq);
+        auto name_len = eq - str;
+        auto name = strndup(str, name_len);
+        _putenv_s(name, "");
+        free(name);
```
Here I create a null-terminated string containing the name only using strndup. I replaced it with less confusing code in [#1935](https://github.com/LadybirdBrowser/ladybird/pull/1935/files#diff-95cb641c88265d7bf8da098bbb93c283da2941e68fc04d6c6a81c1e3789de5adR185-R187).
